### ISSUE TITLE
wgsl: unrecognized diagnostics

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#3927]
Currently the spec says that an unrecognized diagnostic rule is *ignored*.

The motivation for this is:
* allow for deprecation and removal of old diagnostics.
     * E.g. if we decide that derivative_uniformity was a big mistake, we can delete it from the spec, and delete its supporting code in implementations.
* allow implementations to be tolerant to new diagnostic controls.

The downside to this is:
* If you mis-spell a diagnostic control, you'll never hear about it.
     * One answer to this is to catch this in a linter

This is followup to https://github.com/gpuweb/gpuweb/issues/3714 and @jimblandy's comments at https://github.com/gpuweb/gpuweb/pull/3713#discussion_r1120522953 

@jimblandy suggested:
-  for WGSL v1, an unrecognized diagnostic_rule_name is an error.
- beyond v1, consider more flexible naming schemes, like Rust's warning scheme.  If I understand correctly, diagnostic rules can have namespaces, and namespaced rules are ignored if unknown.

**Comment by dneto0** [comment]
Quick thought.

We could support directive `diagnostic(strict);`  or similar,  but only as a directive (at the top of the file).
That's lighterweight than #3714.

It reminds me of Perl's `use strict;` which tightens a bunch of things including variable spelling/ declaration requirements.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-03-07-minutes#wgsl-unrecognized-diagnostics-3927">WGSL 2023-03-07 Minutes</a></summary>

* DN: Currently, unrecognized is ignored for diag names. The bad part is that if you fat-finger it and misspell, you don’t get the effect you wanted. And that sounds bad.
* DN: JB proposed making it an error, to be forward-compat and flexible, and also safer.  Don’t have to deal with the flexibility a future design discussion.
* JB: Yeah we don’t need to do anything more than make this an error for now. I have examples of how rust approaches this too, but we don’t need it for v1. It’s important to me to leave the way open. We want to start with strict and relax later.
* AB: We already have diagnostics [in code] and if you run it in e.g. Firefox it will fail, which is a bad experience. I also might recommend a diagnostic-diagnostic. 
* JB: We would rather deal with things by adding recognition of the diagnostic to Firefox, instead of trying to prevent it ahead of time.
* JB:  Whenever the rule is “ignore”, it’s a mess. E.g. variables in Makefiles.
* MM: Ignoring things you don’t understand is how things are done in CSS. The best-practice is to e.g. put the syntax that is ideal/expected at the top, and then amend that later with workarounds for partial implementations in browsers.
* MM: I also wanted to ask, I didn’t expect diagnostics to be Productions in the grammar, that e.g. authors type. I expected them to be spec-only things.
* AB: This is where we get our uniformity opt-out.
* MM: I don’t see a rule of the list of things that authors can type.
* AB: Two things, 1) severity, 2) any-string (the diagnostic name)
* DN: [https://www.w3.org/TR/WGSL/#filterable-triggering-rules](https://www.w3.org/TR/WGSL/#filterable-triggering-rules)   is the list of standard names. There is only one right now. ‘Derviative_uniformity’.
* MM: So this is if I want the “Myles” diagnostic to be an error, but no one recognizes that name for this.
* AB: Want to support use cases where people get creative with linter rules, and stuff their source code with it.  Don’t want other implementations to have to add every name in the world.
* JB: Rust’s solution here is to have a hierarchical name;  namespaced names would be ignored. Standard names that don’t have their namespace must be recognized.
* MM: It sounds like then that kind of namespaceing would be a requirement for this feature
* JB: I would love to write that proposal, but not for v1
* MM: Another way to do this would be to not use this same mechanism for offline tooling scenarios. Instead those should be controlled via command line arguments, or the like.
* KG: **want to revisit with more time**. Want to prioritize resolving for V1.

</details>

**Comment by jimblandy** [comment]
I've filed my proposal as a new issue, https://github.com/gpuweb/gpuweb/issues/3989, since it's not as short as I had hoped, and I don't want to make people scroll down into the midst of some other discussion to find it.

There was expressed earlier a "lack of appetite" for this wrinkle, which I think is completely reasonable. However, in committee we did not have a consensus on a minimal acceptable way to v1. Almost all of the language in #3989 is rationale and comparison; I'm hoping that the actually change to the spec is small enough to be digestible as the price of consensus.
